pm: device: move device busy APIs to pm subsystem

gmarull opened the pull request:
Summary:

```
pm: device: move device busy APIs to pm subsystem

The following device busy APIs:

- device_busy_set()
- device_busy_clear()
- device_busy_check()
- device_any_busy_check()

were used for device PM, so they have been moved to the pm subsystem.
This means they are now prefixed with `pm_` and are defined in
`pm/device.h`.

If device PM is not enabled dummy functions are now provided that do
nothing or return `-ENOSYS`, meaning that the functionality is not
available.
```
```
tests: kernel: device: remove redundant busy set and clear

The test_dummy_device test has nothing to do with device busy testing,
so remove the calls to pm_device_busy_set/pm_device_busy_clear.
```
```
pm: device: improve usage of the atomic flags field

- Use an enum for the available flags
- Use ATOMIC_DEFINE macro to declare the flag bit field
- Improve naming
```

gmarull committed:
pm: device: move device busy APIs to pm subsystem

The following device busy APIs:

- device_busy_set()
- device_busy_clear()
- device_busy_check()
- device_any_busy_check()

were used for device PM, so they have been moved to the pm subsystem.
This means they are now prefixed with `pm_` and are defined in
`pm/device.h`.

If device PM is not enabled dummy functions are now provided that do
nothing or return `-ENOSYS`, meaning that the functionality is not
available.

Signed-off-by: Gerard Marull-Paretas <gerard.marull@nordicsemi.no>

nashif reviewed (changes requested):
Need to consider/agree if deprecation is needed here before we move forward with this.

tbursztyka reviewed (approved):
Aside from @nashif's comment, lgtm.

gmarull committed:
tests: kernel: device: remove redundant busy set and clear

The test_dummy_device test has nothing to do with device busy testing,
so remove the calls to pm_device_busy_set/pm_device_busy_clear.

Signed-off-by: Gerard Marull-Paretas <gerard.marull@nordicsemi.no>

gmarull committed:
pm: device: improve usage of the atomic flags field

- Use an enum for the available flags
- Use ATOMIC_DEFINE macro to declare the flag bit field
- Improve naming

Signed-off-by: Gerard Marull-Paretas <gerard.marull@nordicsemi.no>

gmarull commented:
> Need to consider/agree if deprecation is needed here before we move forward with this.

@nashif After a quick discussion yesterday, we think these calls should be considered part of PM (proof is that they use the PM atomic bit field to operate). Since PM API is not stable, deprecation should not be considered for this case.

ceolin commented on subsys/pm/device.c:
```diff
@@ -149,3 +149,39 @@ int pm_device_state_get(const struct device *dev,
+	return 0;
+}
+
+int pm_device_busy_check(const struct device *dev)
```
would be better and more aligned with the rest of the API if it was `bool pm_device_is_busy()`

gmarull replied:
agreed, adjusted it (also the `any` variant). I have also updated release notes with the changes.

gmarull replied:
well, I've checked it again: the functions return `int`, not `bool`, so I'm not sure about the change in that case. I'd only use `_is_` pattern for boolean checks. Adding DNM until this point is clarified.

gmarull replied:
Missed that suggestion included move to `bool`, so it should be ready now.

gmarull committed:
pm: adjust busy check API call return types and naming

Busy check APIs now return boolean type. Due to that change, the
function names have also been adjusted. The common name pattern for
boolean check type APIs is "PREFIX_is_CONDITION". For example,
"pm_device_is_busy".  pm_device_busy_check has been renamed to
pm_device_is_busy and pm_device_any_busy_check to pm_device_is_any_busy.

Signed-off-by: Gerard Marull-Paretas <gerard.marull@nordicsemi.no>

gmarull committed:
doc: release: add details on device busy check API changes

Explain device busy API name changes and provide an equivalence list to
ease migration.

Signed-off-by: Gerard Marull-Paretas <gerard.marull@nordicsemi.no>

gmarull committed:
pm: adjust function prototype argument name

The argument name did not match with the name found in the definition.

Signed-off-by: Gerard Marull-Paretas <gerard.marull@nordicsemi.no>

gmarull committed:
pm: device: add deprecated busy APIs

Add old APIs marked as deprecated. The API keep the same behavior as
before, so external users have time to migrate their codebase.

Signed-off-by: Gerard Marull-Paretas <gerard.marull@nordicsemi.no>

gmarull commented:
@nashif should be ready now
